docs/proposals/20191017-control-plane.md

```diff
@@ -0,0 +1,576 @@
+)
+```
+
+With the following validations:
+
+- If `KubeadmControlPlane.Spec.KubeadmConfigSpec` does not define external etcd (webhook):
+  - `KubeadmControlPlane.Spec.Replicas` is an odd number.
+  - Configuration of external etcd is determined by introspecting the provided `KubeadmConfigSpec`.
+- `KubeadmControlPlane.Spec.Replicas` is >= 0 or is nil
+- `KubeadmControlPlane.Spec.Version != ""` (openapi)
+- `KubeadmControlPlane.Spec.KubeadmConfigSpec` must be treated as immutable (via webhook)
+
+And the following defaulting:
+
+- `KubeadmControlPlane.Spec.Replicas: 1` (openapi or webhook)
+
+#### Modifications required to existing API Types
+
+- Add `Cluster.Spec.ControlPlaneRef` defined as:
+
+```go
+    // ControlPlaneRef is an optional reference to a provider-specific resource that holds
+    // the details for provisioning the Control Plane for a Cluster.
+    // +optional
+    ControlPlaneRef *corev1.ObjectReference `json:"controlPlaneRef,omitempty"`
+```
+
+- Add `Cluster.Status.ControlPlaneReady` defined as:
+
+```go
+    // ControlPlaneReady defines if the control plane is ready.
+    // +optional
+    ControlPlaneReady bool `json:"controlPlaneReady,omitempty"`
+```
+
+#### Behavioral Changes from v1alpha2
+
+- If Cluster.Spec.ControlPlaneRef is set:
+  - [Status.ControlPlaneInitialized](https://github.com/kubernetes-sigs/cluster-api/issues/1243) is set based on the value of Status.Initialized for the referenced resource.
+  - Status.ControlPlaneReady is set based on the value of Status.Ready for the referenced resource, this field is intended to eventually replace Status.ControlPlaneInitialized as a field that will be kept up to date instead of set only once.
+- Current behavior will be preserved if `Cluster.Spec.ControlPlaneRef` is not set.
+- CA certificate secrets that were previously generated by the Kubeadm bootstrapper will now be generated by the KubeadmControlPlane Controller, maintaining backwards compatibility with the previous behavior if the KubeadmControlPlane is not used.
+- The kubeconfig secret that was previously created by the Machine Controller will now be generated by the KubeadmControlPlane Controller, maintaining backwards compatibility with the previous behavior if the KubeadmControlPlane is not used.
+
+#### Behaviors
+
+##### Create
+
+- After a KubeadmControlPlane object is created, it must bootstrap a control plane with a given number of replicas.
+- If an error occurs, `KubeadmControlPlane.Status.ErrorStatus` and `KubeadmControlPlane.Status.ErrorMessage` are populated.
+- `KubeadmControlPlane.Spec.Replicas` must be an odd number.
+- Can create an arbitrary number of control planes if etcd is external to the control plane, which will be determined by introspecting `KubeadmControlPlane.Spec.KubeadmConfigSpec`.
+- Creating a KubeadmControlPlane with > 1 replicas is equivalent to creating a KubeadmControlPlane with 1 replica followed by scaling the KubeadmControlPlane to 3 replicas
```

Comment:
I assume you mean equivalent to create it with 1 replica and scaling it up to the specified count (not necessarily 3)

Reply:
Good catch, thanks!

Reply:
Fixed, thanks!